### PR TITLE
Parmest update of util convert_params_to_vars.

### PR DESCRIPTION
**Pull request by smartin71**
## Summary/Motivation:
This is step 1 of a set of internal changes to parmest related to issue:
https://github.com/Pyomo/pyomo/issues/3252

## Changes proposed in this PR:
- Revamped util function convert_params_to_vars to avoid the use of parameter names as strings.
- Now using component IDs both internally to the convert function and in the API.
-

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/contrib/parmest/utils/model_utils.py [hunk 1]
```diff
@@ -43,21 +43,31 @@ def convert_params_to_vars(model, param_names=None, fix_vars=False):
...
+        param_CUIDs = [
+            ComponentUID(param) for param in model.component_data_objects(pyo.Param)
+        ]
 
-    indexed_param_names = []
+    # if param_CUIDs is None:
+    #     param_CUIDs = [
+    #         ComponentUID(param.name)
+    #         for param in model.component_data_objects(pyo.Param)
+    #     ]
```
Comment (mrmundt) [inline pyomo/contrib/parmest/utils/model_utils.py:55]: Leftover code?

Comment (smartin71) [reply pyomo/contrib/parmest/utils/model_utils.py]: Oops!  Yes, leftover code.

**Review by mrmundt — APPROVED** [review]

File: pyomo/contrib/parmest/utils/model_utils.py [hunk 2]
```diff
@@ -69,20 +73,24 @@ def convert_params_to_vars(model, param_names=None, fix_vars=False):
...
-                param_theta_objects.append(theta_cuid.find_component_on(model))
-                indexed_param_names.append(indexed_param_name)
 
+            # get indexed Params
+            param_theta_objects = [theta_obj for _, theta_obj in theta_object.items()]
+
+            # get indexed Param names
+            indexed_param_CUIDs += [
+                ComponentUID(theta_obj) for _, theta_obj in theta_object.items()
+            ]
```
Comment (jsiirola) [inline pyomo/contrib/parmest/utils/model_utils.py:88]: It would be more clear / more efficient to use list's `extend()` and dict's `values()`:
```python
indexed_param_CUIDs.extend(
    ComponentUID(theta_obj) for theta_obj in theta_object.values()
)
```

File: pyomo/contrib/parmest/utils/model_utils.py [hunk 3]
```diff
@@ -94,13 +102,9 @@ def convert_params_to_vars(model, param_names=None, fix_vars=False):
...
-            var_theta_objects = []
-            for theta_obj in theta_var_object:
-                theta_cuid = ComponentUID(
-                    theta_var_object.name + '[' + str(theta_obj) + ']'
-                )
-                var_theta_objects.append(theta_cuid.find_component_on(model))
-
+            var_theta_objects = [
+                var_theta_obj for _, var_theta_obj in theta_var_object.items()
+            ]
```
Comment (jsiirola) [inline pyomo/contrib/parmest/utils/model_utils.py:107]: If you are only using the values from a dict, it is more clear / more efficient to use `values()`:
```python
var_theta_objects = list(var_theta_obj.values())
```

File: pyomo/contrib/parmest/utils/model_utils.py [hunk 4]
```diff
@@ -124,14 +128,14 @@ def convert_params_to_vars(model, param_names=None, fix_vars=False):
...
-        param_names = indexed_param_names
+    # Update the list of param_CUIDs if the parameters were indexed
+    if len(indexed_param_CUIDs) > 0:
+        param_CUIDs = indexed_param_CUIDs
 
     # Convert Params to Vars in Expressions
     for expr in model.component_data_objects(pyo.Expression):
         if expr.active and any(
-            v.name in param_names for v in identify_mutable_parameters(expr)
+            ComponentUID(v) in param_CUIDs for v in identify_mutable_parameters(expr)
```
Comment (jsiirola) [inline pyomo/contrib/parmest/utils/model_utils.py:138]: This is *very* inefficient (quadratic in the number of params).  If you are going to test membership, `param_CUIDs` needs to be a set / dict / mapping object (so that lookups are O(1) and not O(n)).

**Review by jsiirola — CHANGES_REQUESTED** [review]
This looks pretty good, but there is a big performance bug that should be addressed.

File: pyomo/contrib/parmest/utils/model_utils.py (hunk 2 above)
Comment (smartin71) [reply pyomo/contrib/parmest/utils/model_utils.py]: Changed as suggested.

File: pyomo/contrib/parmest/utils/model_utils.py (hunk 3 above)
Comment (smartin71) [reply pyomo/contrib/parmest/utils/model_utils.py]: Changed as suggested.

File: pyomo/contrib/parmest/utils/model_utils.py (hunk 4 above)
Comment (smartin71) [reply pyomo/contrib/parmest/utils/model_utils.py]: Converted list to set for the lookups at the end of the function.

File: pyomo/contrib/parmest/utils/model_utils.py [hunk 5]
```diff
@@ -43,21 +43,25 @@ def convert_params_to_vars(model, param_names=None, fix_vars=False):
...
-        theta_cuid = ComponentUID(param_name)
-        theta_object = theta_cuid.find_component_on(model)
+        theta_object = param_CUID.find_component_on(model)
 
         # Param
         if theta_object.is_parameter_type():
+
             # Delete Param, add Var
             vals = theta_object.extract_values()
             model.del_component(theta_object)
```
Comment (blnicho) [inline pyomo/contrib/parmest/utils/model_utils.py:67]: I don't think this will work on hierarchical models. I think we need to add more tests for edge cases.

File: pyomo/contrib/parmest/utils/model_utils.py [hunk 6]
```diff
@@ -124,14 +126,18 @@ def convert_params_to_vars(model, param_names=None, fix_vars=False):
     if len(substitution_map) == 0:
         return model
 
-    # Update the list of param_names if the parameters were indexed
-    if len(indexed_param_names) > 0:
-        param_names = indexed_param_names
+    # Update the list of param_CUIDs if the parameters were indexed
+    if len(indexed_param_CUIDs) > 0:
+        param_CUIDs = indexed_param_CUIDs
```
Comment (blnicho) [inline pyomo/contrib/parmest/utils/model_utils.py:131]: Why do we replace `param_CUIDs` here instead of extending it?

File: pyomo/contrib/parmest/utils/model_utils.py [hunk 7]
```diff
@@ -69,20 +73,24 @@ def convert_params_to_vars(model, param_names=None, fix_vars=False):
...
-            for theta_obj in theta_object:
-                indexed_param_name = theta_object.name + '[' + str(theta_obj) + ']'
-                theta_cuid = ComponentUID(indexed_param_name)
-                param_theta_objects.append(theta_cuid.find_component_on(model))
-                indexed_param_names.append(indexed_param_name)
 
+            # get indexed Params
+            param_theta_objects = [theta_obj for _, theta_obj in theta_object.items()]
+
+            # get indexed Param names
```
Comment (blnicho) [inline pyomo/contrib/parmest/utils/model_utils.py:85]:
```suggestion
            # get indexed Param CUIDs
```

**Review by blnicho — CHANGES_REQUESTED** [review]
I think this still needs some work. There are some bugs in this function when params to be converted to vars live on blocks. This PR didn't necessarily introduce the bugs but I think we should fix them while we're overhauling this function. Here is a small example demonstrating one of the bugs:

```
from pyomo.environ import *
from pyomo.contrib.parmest.utils import convert_params_to_vars

m = ConcreteModel()
m.p1 = Param(initialize=1, mutable=True)
m.b = Block()
m.b.p2 = Param(initialize=2, mutable=True)

param_cuids = [ComponentUID(m.p1), ComponentUID(m.b.p2)]
m2 = convert_params_to_vars(m, param_cuids)
```

this fails with an `AttributeError`:
`AttributeError: 'ScalarParam' object has no attribute 'unfix'`

File: pyomo/contrib/parmest/parmest.py [hunk 8]
```diff
@@ -440,8 +440,8 @@ def TotalCost_rule(model):
             )
 
         # Convert theta Params to Vars, and unfix theta Vars
-        theta_names = [k.name for k, v in model.unknown_parameters.items()]
-        parmest_model = utils.convert_params_to_vars(model, theta_names, fix_vars=False)
+        theta_CUIDs = [v for k, v in model.unknown_parameters.items()]
```
[the remark this answers is not in the document]
Comment (smartin71) [reply pyomo/contrib/parmest/parmest.py:443]: Changed.

File: pyomo/contrib/parmest/tests/test_utils.py [hunk 9]
```diff
@@ -46,12 +50,12 @@ def test_convert_param_to_var(self):
         exp = ReactorDesignExperiment(data, 0)
         instance = exp.get_labeled_model()
 
-        theta_names = ['k1', 'k2', 'k3']
+        param_CUIDs = [v for k, v in instance.unknown_parameters.items()]
```
[the remark this answers is not in the document]
Comment (smartin71) [reply pyomo/contrib/parmest/tests/test_utils.py:53]: Changed.

File: pyomo/contrib/parmest/utils/model_utils.py [hunk 10]
```diff
@@ -69,20 +73,24 @@ def convert_params_to_vars(model, param_names=None, fix_vars=False):
...
             vals = theta_object.extract_values()
-            param_theta_objects = []
-            for theta_obj in theta_object:
-                indexed_param_name = theta_object.name + '[' + str(theta_obj) + ']'
-                theta_cuid = ComponentUID(indexed_param_name)
-                param_theta_objects.append(theta_cuid.find_component_on(model))
-                indexed_param_names.append(indexed_param_name)
 
+            # get indexed Params
+            param_theta_objects = [theta_obj for _, theta_obj in theta_object.items()]
```
[the remark this answers is not in the document]
Comment (smartin71) [reply pyomo/contrib/parmest/utils/model_utils.py:83]: Changed.

File: pyomo/contrib/parmest/utils/model_utils.py [hunk 11]
```diff
@@ -124,14 +126,18 @@ def convert_params_to_vars(model, param_names=None, fix_vars=False):
...
+    # Update the list of param_CUIDs if the parameters were indexed
+    if len(indexed_param_CUIDs) > 0:
+        param_CUIDs = indexed_param_CUIDs
+
+    # convert to a set for look up efficiency
+    param_CUIDs_set = set(param_CUIDs)
 
     # Convert Params to Vars in Expressions
     for expr in model.component_data_objects(pyo.Expression):
         if expr.active and any(
```
[the remark this answers is not in the document]
Comment (smartin71) [reply pyomo/contrib/parmest/utils/model_utils.py:138]: Changed.

File: pyomo/contrib/parmest/utils/model_utils.py (hunk 6 above)
Comment (smartin71) [reply pyomo/contrib/parmest/utils/model_utils.py:131]: I think we have to replace it because it might not have the indexed variables, only the main variable.  For example "theta" (param_CUIDs) vs. "theta[asymptote]" and "theta[rate_constant]" (indexed_param_CUIDs).

File: pyomo/contrib/parmest/utils/model_utils.py (hunk 7 above)
Comment (smartin71) [reply pyomo/contrib/parmest/utils/model_utils.py:85]: Changed.

File: pyomo/contrib/parmest/utils/model_utils.py (hunk 5 above)
Comment (smartin71) [reply pyomo/contrib/parmest/utils/model_utils.py]: I think I fixed this and added a test.

**Comment by mrmundt** [comment]
@blnicho - Do you think you'll have time to look over this one before the Monday patch?

**Comment by shermanjasonaf** [comment]
Does this fix #3375?

**Comment by adowling2** [comment]
@djlaky @smondal13 FYI, this is related to some of our Pyomo.DoE development plans

File: pyomo/contrib/parmest/utils/model_utils.py [hunk 12]
```diff
@@ -16,22 +16,23 @@
...
 from pyomo.common.collections import ComponentMap
+from pyomo.core.base import Model
 
 from pyomo.environ import ComponentUID
 
 logger = logging.getLogger(__name__)
 
 
-def convert_params_to_vars(model, param_names=None, fix_vars=False):
+def convert_params_to_vars(model, param_CUIDs=None, fix_vars=False):
```
Comment (blnicho) [inline pyomo/contrib/parmest/utils/model_utils.py:26]: TODO: make sure the param domain is also being transferred to the new var

**Comment by adowling2** [comment]
Is this something that @slilonfe5 or others at Notre Dame should take over? Or is this not needed, and the PR should get closed? If we are taking over the PR, what is the best course of action? Should @smartin71 make this branch editable by others? Or should @slilonfe5 create a new branch based on this current branch?
